### PR TITLE
NF: rudimentary web interface

### PR DESCRIPTION
**Pull request by debanjum**
yoh: supersedes original https://github.com/datalad/datalad/pull/541 
since over 100 commits, buildbot tests were ran in #759 and passed

### CODE REVIEW

**Comment by yarikoptic** [comment]
@datalad/developers  you have unique opportunity of 24h to review/complaint or better extend (send PRs to @debanjum) this branch or we would get it merged tomorrow and use to provide an improved browsing of datasets.  For demo, visit  http://datasets.datalad.org/webtest/

**Comment by aqw** [comment]
@yarikoptic That's a big PR to review. ;-) I'll try to take a look at the HTML/CSS/JS early tomorrow (CET), but I expect any comments I have will be best provided as an independent, later PR.

---Alex

**Comment by yarikoptic** [comment]
Thank you @aqw in advance. Indeed PR is somewhat sizeable although majority is just 3rd party pieces, but yes, there is dune python and JavaScript ( with tests) to consider.

File: datalad/resources/website/index.html [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    </head>
+    <body>
+        <h1><img src='.git/datalad/web/assets/images/datalad_banner.svg' width="196"></h1>
+        <table id="directory" class="display" width="100%" cellspacing="0"></table>
+	<noscript>
+	    <div style="border: 2px solid #E53B51; padding: 10px">
+		<span style="font-weight:bold">Javascript is not enabled!</span>
+	    </div>
+	</noscript>
+	<script type="text/javascript" charset="utf8" src=".git/datalad/web/assets/js/jquery-1.12.3.js"></script>
```
Comment (aqw) [inline datalad/resources/website/index.html]: These should be `charset="utf-8"` to validate properly.

File: datalad/resources/website/index.html [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+         td.dir {font-weight: bold; color: #00A1CB; }
+         td.git {font-weight: bold; color: #00BA88; }
+         td.annex {font-weight: bold; color: #EC6C20; }
+         td.link-broken {color:#E53B51; }
+         td.link {color: #FFD03C;  }
+         td.file {color: #67CDDC;  }
+        </style>
+    </head>
+    <body>
+        <h1><img src='.git/datalad/web/assets/images/datalad_banner.svg' width="196"></h1>
```
Comment (aqw) [inline datalad/resources/website/index.html]: The `` needs the `alt` attribute, even more-so since you're playing the edge of semantics/accessibility by having no content in the `<h1>`.

File: datalad/resources/website/index.html [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        <table id="directory" class="display" width="100%" cellspacing="0"></table>
```
Comment (aqw) [inline datalad/resources/website/index.html]: Neither `width` nor `cellspacing` are valid HTML5 anymore, and should be removed. It seems your CSS already covers that anyway.

File: datalad/resources/website/index.html [hunk 4]
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <meta charset="UTF-8">
+        <title>Datalad Repository</title>
+        <link rel="shortcut icon" type="image/x-icon" href=".git/datalad/web/assets/images/favicon-16x16.png" />
+        <link rel="stylesheet" type="text/css" href=".git/datalad/web/assets/css/jquery.dataTables.css">
+        <style type="text/css" class="init">
```
Comment (aqw) [inline datalad/resources/website/index.html]: What does `class="init"` do here? My google-foo is failing me, and I can't see a clear purpose for it.

File: datalad/resources/website/assets/js/main.js [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+ */
+function update_param_or_path(next_url, type, current_state) {
+  // if url = root path(wrt index.html) then append index.html to url
+  // allows non-root dataset dirs to have index.html
+  // ease constrain on non-datalad index.html presence in dataset
+  if (next_url === loc().pathname || next_url === '/' || !next_url) {
+    return false;
+  } else if (type === 'file' || type === 'link')
+    return false;
+  else
```
Comment (aqw) [inline datalad/resources/website/assets/js/main.js]: Just a style comment, feel free to ignore. But I've never liked the shortcut of omitting of curly braces for one-line if/elses in JS. And here you're mixing both styles side-by-side. It visually implies the flow is different, even though, to my eye, it isn't.

**Comment by aqw** [comment]
@debanjum Just got done taking a look at the web bits of the PR. Overall, it looks good. Most of my comments came from simply running the page through the HTML validator.

Is there any reason in particular that you're choosing to use jQuery 1.x? I know it's still supported, but it seems like an odd choice for a new project. Data Tables works with both 1.x and 2.x. The only difference I can see is for IE6-8 support, but if that's the case, I'm all for bumping to 2.x, dropping those browsers, and saving ~10KB in the process. The next version in 1.x drops < IE8 anyway, so it seems there's little to gain here.

jQuery 1.x isn't a show stopper for me, but it does merit consideration.

@yarikoptic Once the small changes I suggested are addressed, it LGTM.

---Alex

**Comment by debanjum** [comment]
Hey, thanks a ton for the review of the HTML and JS stuff! It really needed it ;)
I'll enable some kind of HTML validator on my setup to help out. The JS linter proved really helpful while coding the js up.

No particular reason that I can remember of for using jQuery 1.x so i'll move it to 2.x and hopefully everything should work fine.

File: datalad/resources/website/index.html (hunk 4 above)
Comment (debanjum) [inline datalad/resources/website/index.html]: I'm guessing this allows datatable to override its css with this custom one in case of conflict or some such, not sure though

File: datalad/resources/website/index.html (hunk 3 above)
Comment (debanjum) [inline datalad/resources/website/index.html]: that's interesting, I had no clue.

Comment (aqw) [inline datalad/resources/website/index.html]: Yeah, pretty much any HTML that is styling only is (or is in the process of being) deprecated in favor of CSS. That is, except for the raging debate around `<hr>`. ;-)

---Alex

**Comment by aqw** [comment]
@debanjum Glad I could be of some help.

Nice work!

---Alex

File: datalad/resources/website/index.html (hunk 4 above)
Comment (aqw) [inline datalad/resources/website/index.html]: Overwriting (vs merely overriding) CSS in an `<style>` tag and acquiring that element via a class name would be some pretty messed-up-foo IMO.

If I were to guess, this is either an accidental copy-paste or a leftover from a clever/insane approach to providing an interactive example.

In either case, I'd recommend removing it, as I was surprised it's even valid (and I think it only is by accident, by way of `class` being a global attribute. The specs (4, 5, XHTML, nor the Living Standard) make no mention of any semantic difference or purpose to using `class` on a `<style>` tag).

If everything breaks, obviously keep it, and please let me know, as I will be /very/ interested.

---Alex

Comment (debanjum) [inline datalad/resources/website/index.html]: Yeah, don't know where I got that snippet/idea from. The only place I could find a reference to using class="init" in datatables is [here](https://github.com/DataTables/DataTables/blob/master/examples/server_side/ids.html). 
Anyway removing it seems to be working fine, so fixed it. 
Better to remove it then to have unknown cruft ;)

File: datalad/resources/website/index.html (hunk 3 above)
Comment (debanjum) [inline datalad/resources/website/index.html]: That would be great! Will make the html a lot cleaner

**Comment by yarikoptic** [comment]
please merge current master and push! watchout for conflicts

File: datalad/distribution/create_publication_target_sshwebserver.py [hunk 6]
```diff
@@ -232,97 +242,188 @@ def __call__(sshurl, target=None, target_dir=None,
...
+
+        # create json command for current dataset
+        json_command = 'which datalad > /dev/null && (cd ..; GIT_DIR=$PWD/.git datalad ls -r --json file \'{}\';) &> "{}/{}"'
+        json_command = json_command.format(str(path), logs_remote_target, 'datalad-publish-hook-$(date +%Y-%m-%dT%H:%M:%S%z).log')
+
+        # collate content for post_update hook
+        hook_content = '\n'.join(['#!/bin/bash', 'git update-server-info', json_command])
+
+        # make datalad logs directory
+        ssh(['mkdir', '-p', logs_remote_target])
```
Comment (yarikoptic) [inline datalad/distribution/create_publication_target_sshwebserver.py]: You could've added this to hook itself... Make a separate PR since this I've hopefully will get merged!:-)

File: datalad/distribution/create_publication_target_sshwebserver.py [hunk 7]
```diff
@@ -232,97 +242,188 @@ def __call__(sshurl, target=None, target_dir=None,
...
+                "Failed to determine git version on remote.\n"
+                "Error: {0}\nTrying to configure anyway "
+                "...".format(exc_str(e)))
+        return None
+
+    @staticmethod
+    def create_postupdate_hook(path, ssh, dataset):
+        # location of post-update hook file, logs folder on remote target
+        hook_remote_target = opj(path, '.git', 'hooks', 'post-update')
+        logs_remote_target = opj(path, '.git', 'datalad', 'logs')
```
Comment (yarikoptic) [inline datalad/distribution/create_publication_target_sshwebserver.py]: For any datalad paths, better define them in consts.py we have so we could see all the paths we care about in one place
